util/stream.go

```diff
@@ -156,40 +136,76 @@ func (m *MessageStream) inbound() {
 			return
 		}
 
-		for i := 0; i < n; i++ {
-			if hdr < 4 {
-				hdrBuf[hdr] = tmpBuf[i]
-				buf.WriteByte(tmpBuf[i])
-				hdr += 1
-				if hdr >= 4 {
-					// MessageStream is not protocol agnostic. Reading length based
-					// on OpenFlow header field.
-					msgLen = int(binary.BigEndian.Uint16(hdrBuf[2:])) - 4
+		// Append the bytes read from the connection to buf.
+		buf.Write(tmpBuf[:n])
+
+		// Read from the connection until the OpenFlow message header is retrieved.
+		for buf.Len() >= 4 {
+			if totalLen == 0 {
+				// Read the OpenFlow message length.
+				msgType := int(buf.Bytes()[1])
+				totalLen = int(binary.BigEndian.Uint16(buf.Bytes()[2:4]))
+				// msgType == openflow15.Type_Experimenter
+				if msgType == 4 {
+					// The minimum length of a valid VendorHeader message is 16 bytes.
+					if buf.Len() < 16 {
+						break
+					}
+					experimenterType := binary.BigEndian.Uint32(buf.Bytes()[12:])
+					// experimenterType == openflow15.Type_PacketIn2
+					if experimenterType == 30 {
+						// The first 4 byte of a PacketIn2 message is needed to check the packet length.
+						if buf.Len() < 20 {
+							break
+						}
+						// According to OVS implementation, the first property of a PacketIn2 message is NXPINT_PACKET.
+						pktProp := int(binary.BigEndian.Uint16(buf.Bytes()[16:]))
+						// pkgProp == openflow15.NXPINT_PACKET
+						if pktProp == 0 {
+							pktLength := int(binary.BigEndian.Uint16(buf.Bytes()[18:]))
+							if totalLen < pktLength {
+								totalLen += 1 << 16
+								klog.V(2).InfoS("Oversize packet detected: OpenFlow PacketIn message length overflowed", "message_length", totalLen)
+								// Reset the VendorHeader.Vendor field to mark the message is oversize.
+								binary.BigEndian.PutUint32(buf.Bytes()[8:12], 0x10002320)
+							}
+						}
+					}
 				}
-				continue
-			}
-			if msgLen > 0 {
-				buf.WriteByte(tmpBuf[i])
-				msgLen = msgLen - 1
-				if msgLen == 0 {
-					hdr = 0
-					m.dispatchMessage(buf)
-					buf = <-m.pool.Empty
+				klog.V(5).InfoS("Expected OpenFlow message", "length", totalLen)
```

Comment:
why "Expected"?

Reply:
I could see how it could be "Expecting", as we have determined the length based on header fields but we have yet to read the full message. I could see "Receiving" being an option as well.

Reply:
updated.